[1.19.x] Rewrite Getting Started (#490)
MinecraftForge/Documentation: docs/gettingstarted/structuring.md

Structuring Your Mod
====================

Structured mods are beneficial for maintenance, making contributions, and providing a clearer understanding of the underlying codebase. Some of the recommendations from Java, Minecraft, and Forge are listed below.

!!! note

>     You do not have to follow the advice below; you can structure your mod anyway you set fit. However, it is still highly recommended to do so.

Reviewer: ```suggestion
    You do not have to follow the advice below; you can structure your mod any way you set fit. However, it is still highly recommended to do so.
```

Packaging
---------

When structuring your mod, pick a unique, top-level package structure. Many programmers will use the same name for different classes, interfaces, etc. Java allows classes to have the same name as long as they are in different packages. As such, if two classes have the same package with the same name, only one would be loaded, most likely causing the game to crash.

```
a.jar
  - com.example.ExampleClass
b.jar
  - com.example.ExampleClass // This class will not normally be loaded
```

> This is even more relevant when it comes to loading modules. If there are class files in two packages under the same name in separate modules, then only the first package will be loaded.

Reviewer: This is false, duplicate packages across modules will result in a hard failure at startup, as mod modules are exported to the game and each other.

> This is even more relevant when it comes to loading modules. If there are class files in two packages under the same name in separate modules, this will cause the mod loader to crash on startup since mod modules are exported to the game and other mods.

Reviewer 2: Perhaps emphasis should be added on the "to crash on startup" part of the paragraph.
```suggestion
This is even more relevant when it comes to loading modules. If there are class files in two packages under the same name in separate modules, this will cause the mod loader **to crash on startup** since mod modules are exported to the game and other mods.
```

```
module A
  - package X
    - class I
    - class J
module B

>   - package X // This package will not normally be loaded, preventing usage of R, S, T

Reviewer 2: As a bit of nuance: By my understanding, Java doesn't crash by the mere presence of two modules have the same package. The part that causes the crash is when two modules co-exist and try to export that same package out to another module (or all other modules, in the case of `open` modules).

As Java provides APIs to build module layers programmatically, an application or framework can load multiple modules that have the same package as long as each module does not co-exist with another on the same layer and export the same package.

Author: I'm not sure whether that should be stated, but I fixed the message to reflect the text above mentioning the crash.

    - class R
    - class S
    - class T
```

As such, your top level package should be something that you own: a domain, email address, a subdomain of where your website, etc. It can even be your name or username as long as you can guarantee that it will be uniquely identifiable within the expected target.

Type      | Value             | Top-Level Package
:---:     | :---:             | :---
Domain    | example.com       | `com.example`
Subdomain | example.github.io | `io.github.example`
Email     | example@gmail.com | `com.gmail.example`

> The next level package should then be your mod's id (e.g. `com.example.examplemod` where `examplemod` is the mod id). This will guarantee that, unless you have two mods with the same mod id (which should never be the case), your packages should not have any issues loading.

Reviewer: ```suggestion
The next level package should then be your mod's id (e.g. `com.example.examplemod` where `examplemod` is the mod id). This will guarantee that, unless you have two mods with the same id (which should never be the case), your packages should not have any issues loading.
```

You can find some additional naming conventions on [Oracle's tutorial page][naming].

### Sub-package Organization

> In addition to the top-level package, it is highly recommend to break your mod's classes between subpackages. There are numerous methods on how to do so:

Reviewer 2: Re-word the last sentence, since we only provide two methods here (and usually these are the major two methods that are used):
```suggestion
In addition to the top-level package, it is highly recommend to break your mod's classes between subpackages. There are two major methods on how to do so:
```

> * **Group By Function**: Make subpackages for classes with a common purpose. For example, blocks can be under `block` or `blocks`, entities under `entity` or `entities`, etc. Minecraft uses this structure with the singular version of the word.

Reviewer: Given it is names and we're talking about official mappings.
```suggestion
* **Group By Function**: Make subpackages for classes with a common purpose. For example, blocks can be under `block` or `blocks`, entities under `entity` or `entities`, etc. Mojang uses this structure with the singular version of the word.
```

> * **Group By Logic**: Make subpackages for classes with a common logic. For example, if you were creating a new type of crafting table, you would put its block, menu, item, and more under `feature/crafting_table`.

Reviewer: ```suggestion
* **Group By Logic**: Make subpackages for classes with a common logic. For example, if you were creating a new type of crafting table, you would put its block, menu, item, and more under `feature.crafting_table`.
```

#### Client, Server, and Data Packages

In general, code only for a given side or runtime should be isolated from the other classes in a separate subpackage. For example, code related to [data generation][datagen] should go in a `data` package while code only on the dedicated server should go in a `server` package.

However, it is highly recommended that [client-only code][sides] should be isolated in a `client` subpackage. This is because dedicated servers have no access to any of the client-only packages in Minecraft. As such, having a dedicated package would provide a decent sanity check to verify you are not reaching across sides within your mod.

Class Naming Schemes
--------------------

A common class naming scheme makes it easier to decipher the purpose of the class or to easily locate specific classes.

Classes are commonly suffixed with its type, for example:

* An `Item` called `PowerRing` -> `PowerRingItem`.
* A `Block` called `NotDirt` -> `NotDirtBlock`.
* A menu for an `Oven` -> `OvenMenu`.

!!! note

>     Minecraft typically follows a similar structure for all classes except entities. Those are represented by just their names (e.g. `Pig`, `Zombie`, etc.).

Reviewer: ```suggestion
    Mojang typically follows a similar structure for all classes except entities. Those are represented by just their names (e.g. `Pig`, `Zombie`, etc.).
```

Choose One Method from Many
---------------------------

> There are many methods for performing a certain task: registering an object, listening for events, etc. It's generally recommended to only use a single method to accomplish a given task. While this does improve code formatting, it also avoid any weird interactions or redundancies that may occur (e.g. your event listener executing twice).

Reviewer 2: Perhaps this should mention "consistency" as well; stick to one way to do it and be consistent throughout your code.

Author: It would help if I remembered that consistency was a word when I originally wrote this.
